pool/result_pool.go

```diff
@@ -83,11 +84,57 @@ func (p *ResultPool[T]) panicIfInitialized() {
 // goroutines. The zero value is valid and ready to use.
 type resultAggregator[T any] struct {
 	mu      sync.Mutex
+	len     int
 	results []T
+	errored []int
 }
```

Comment:
The changes to `resultAggregator` are central to this PR.

Summary:
- Add a `nextIndex` method that makes it possible to reserve a slot in the result slice
- Change the `add` method to `save`, which also takes an the index from `nextIndex` and whether the result errored
- Add a `collect` method that respects the `WithCollectErrored` option by filtering out errored results.